### PR TITLE
CSHARP-4700: Fix issue with GroupBy followed by OrderBy.

### PR DESCRIPTION
**Pull request by rstam**

### CODE REVIEW

File: src/MongoDB.Driver/Linq/Linq3Implementation/Ast/Optimizers/AstGroupingPipelineOptimizer.cs [hunk 1]
```diff
@@ -169,11 +170,20 @@ static bool StageCanBeOptimized(AstStage stage)
 
                 static bool IsLastStageThatCanBeOptimized(AstStage stage)
                 {
-                    return stage.NodeType switch
+                    return stage switch
                     {
-                        AstNodeType.ProjectStage => true,
+                        AstProjectStage projectStage => !ProjectsRoot(projectStage),
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Ast/Optimizers/AstGroupingPipelineOptimizer.cs:175]: When optimizing a project following a grouping stage we can't optimize the _elements field away when the project projects the $$ROOT variable. In this case $$ROOT is an indirect reference to the _elements field and we must keep the unoptimized _elements field.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4700Tests.cs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+                AssertStages(
+                    stages,
+                    "{ $group : { _id : '$Name', __agg0 : { $sum : 1 } } }",
+                    "{ $sort : { __agg0 : 1 } }");
+
+                results.Should().HaveCount(2);
+                results[0].Key.Should().Be("Jane");
+                results[0].Count().Should().Be(0); // this result is incorrect in LINQ2
+                results[1].Key.Should().Be("John");
+                results[1].Count().Should().Be(0); // this result is incorrect in LINQ2
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp4700Tests.cs:51]: LINQ2 doesn't throw an exception but the result is wrong. The grouping is missing the elements.

**Review by JamesKovacs — APPROVED** [review]
LGTM

**Review by sanych-sun — APPROVED** [review]
LGTM

**Pull request merged**